Use SearchStats instead of field.isAggregatable in data node planning

craigtaverner opened the pull request:
Since ES|QL makes use of field-caps and only considers `isAggregatable` during Lucene pushdown, turning off doc-values disables Lucene pushdown. This is incorrect. The physical planning decision for Lucene pushdown is made during local planning on the data node, at which point `SearchStats` are known, and both `isIndexed` and `hasDocValues` are separately knowable. The Lucene pushdown should happen for `isIndexed` and not consider `hasDocValues` at all.

This PR adds hasDocValues to SearchStats and the uses isIndexed and hasDocValue separately during local physical planning on the data nodes. This immediately cleared up one issue for spatial data, which could not push down a lucene query when doc-values was disabled.

Summary of what `isAggregatable` means for different implementations of `MappedFieldType`:
* Default implementation of `isAggregatable` in `MappedFieldType` is `hasDocValues`, and does not consider `isIndexed`
* All classes that extend `AbstractScriptFieldType` (eg. `LongScriptFieldType`) hard coded `isAggregatable` to `true`. This presumably means Lucene is happy to mimic having doc-values
* `TestFieldType`, and classes that extend it, return the value of `fielddata`, so consider the field aggregatable if there is field-data.
* `AggregateDoubleMetricFieldType` and `ConstantFieldType` hard coded to `true`
* `DenseVectorFieldType` hard coded to `false`
* `IdFieldType` return the value of `fieldDataEnabled.getAsBoolean()`

In no case is `isIndexed` used for `isAggregatable`. However, for our Lucene pushdown of filters, `isIndexed` would make a lot more sense. But for pushdown of TopN, `hasDocValues` makes more sense.

Summarising the results of the various options for the various field types, where `?` means configrable:

| Class | isAggregatable | isIndexed | isStored | hasDocValues |
| --- | --- | --- | --- | --- |
| AbstractScriptFieldType                 | true  | false | false | false |
| AggregateDoubleMetricFieldType | true  | true  | false | false |
| DenseVectorFieldType                    | false | ?       | false | !indexed |
| IdFieldType                                      | fieldData | true | true | false |
| TsidExtractingIdField                       | false | true | true | false |
| TextFieldType                                   | fieldData | ? | ? | false |
| ? (the rest)                                        | hasDocValues | ? | ? | ? |

It has also been observed that we cannot push filters to source without checking `hasDocValues` when we use the `SingleValueQuery`. So this leads to three groups of conditions:

| Category | require `indexed` | require `docValues` |
| --- | --- | --- |
| Filters(single-value) | true | true |
| Filters(multi-value) | true | false |
| TopN | true | true |

And for all cases we will also consider `isAggregatable` as a disjunction to cover the script field types, leading to two possible combinations:

* `fa.isAggregatable() || searchStats.isIndexed(fa.name()) && searchStats.hasDocValues(fa.name())`
* `fa.isAggregatable() || searchStats.isIndexed(fa.name())`

Fixes #115737

craigtaverner committed:
Use SearchStats instead of field.isAggregatable in data node planning

craigtaverner committed:
Update docs/changelog/115744.yaml

elasticsearchmachine commented:
Hi @craigtaverner, I've created a changelog YAML for you.

craigtaverner committed:
Remove more cases of isAggregatable from physical planning

craigtaverner committed:
Merge branch 'esql_searchstats_for_indexed_docvalues' of github.com:craigtaverner/elasticsearch into esql_searchstats_for_indexed_docvalues

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_searchstats_for_indexed_docvalues

craigtaverner committed:
Fix semantic text planning

craigtaverner committed:
Fix failing union-types tests

Don't use original field name, but attribute name, so we don't inadvertently check synthetic attributes for indexed status using their source attribute names.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/stats/SearchStats.java:
```diff
@@ -252,6 +254,26 @@ public boolean isIndexed(String field) {
+        var stat = cache.computeIfAbsent(field, s -> new FieldStat());
+        if (stat.hasDocValues == null) {
+            stat.hasDocValues = false;
+            if (exists(field)) {
+                boolean hasDocValues = true;
+                for (SearchExecutionContext context : contexts) {
+                    if (context.isFieldMapped(field)) {
+                        if (context.getFieldType(field).hasDocValues() == false) {
+                            hasDocValues = false;
+                            break;
+                        }
+                    }
```
Since exists() doesn't return the information fully, since it's the same code underneath, how about copying its implementation inside docValues and performing the hasDocValues() check as well - so there's only one SearchExecutionContext loop and the same context is called twice?

craigtaverner replied:
I agree. This pattern of doing the loop for each parameter exists for a few pre-existing methods, so I kept it that way, but probably it would be better to instead capture them all in one go. I think no-one cared before because this check is not in performance critical runtime code, but in some planning code. By merging loops we would primarily share the `isFIeldMapped` and `getFieldType` calls on the `SearchExecutionContext`. All the rest if different between loops.

craigtaverner replied:
OK. I separated the FieldStat into two groups, things that are simple config settings, like `exists`, `indexed`, `hasDocValues` and `hasIdenticalDelegate`, and other things that might take longer to retrieve. The first set are in a record created in a single pass loop (as you suggested). The rest are accessed on demand. I also made refactoring less flaky by not having all test versions extend the production version. I also deleted the entire unused methods for getting `min` and `max`, and think we should only add those back if (hopefully when) we use them.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/stats/SearchStats.java:
```diff
@@ -51,6 +51,7 @@ private static class FieldStat {
         private Boolean singleValue;
         private Boolean hasIdenticalDelegate;
         private Boolean indexed;
+        private Boolean hasDocValues;
```
looks like the naming convention is to have no has prefix so `docValues`

craigtaverner replied:
Or add `is` and `has` prefixes to the rest... (which might conform better to the naming contexts elsewhere).

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushTopNToSource.java:
```diff
@@ -60,7 +62,9 @@
     @Override
     protected PhysicalPlan rule(TopNExec topNExec, LocalPhysicalOptimizerContext ctx) {
-        Pushable pushable = evaluatePushable(topNExec, x -> LucenePushDownUtils.hasIdenticalDelegate(x, ctx.searchStats()));
+        Predicate<FieldAttribute> hasIdenticalDelegate = (fa) -> LucenePushDownUtils.hasIdenticalDelegate(fa, ctx.searchStats());
```
would be useful to rename this method to something more meaningful such as `hasExactFallback`.

craigtaverner replied:
While this method pre-dates my work (was added almost a year ago), I agree that the name is not entirely clear, and my last commit also changed slightly what is checked deeper down too. However, the name `hasExactFallback` is also insufficient (and to my ears sounds synonymous with what we already have). We check if the TEXT field-type has a surrogate and then make several checks on the nature of that surrogate. Any name that fully captures what is checked will be unreasonably long, which is why I kept the original name coded before.

costin reviewed (commented):
👍 
It's a good direction and having this information locally is good.
Please state FTR, what turns on/off isAggregateable in field_caps as there might be other cases where this information is not useful.

Secondly, please look at the impact of removing isAggregatable from EsField.

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_searchstats_for_indexed_docvalues

craigtaverner committed:
Differentiate between isIndexed and isStored for TEXT field delegates

craigtaverner committed:
Require doc-values for pushing down TopN

Not sure why this is required, but let's re-evaluate later. We needed this to get a test to pass.

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_searchstats_for_indexed_docvalues

craigtaverner committed:
Also consider isAggregatable for non-indexed fields like ScriptField

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_searchstats_for_indexed_docvalues

craigtaverner committed:
Refactor SearchStats to collect all config booleans in one loop

Also fixed a bug with hasIdenticalDelegate which returned true if the field did not exist. This required a matching fix to the TestSearchStats.

craigtaverner committed:
Fixed CsvTests which need to mock SearchStats properly

craigtaverner committed:
Fix NPE for PushStatsToSource

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_searchstats_for_indexed_docvalues

craigtaverner committed:
Refactored SearchStats to interface, to reduce subtle bugs in tests

The older approach with test implementations of SearchStats extending the original with an empty list lead to subtle bugs when adding more methods to SearchStats and not also adding overriding methods to the various test implementations.
This refactoring forces test implementations to always implement their own versions of all methods. There was, in fact, no test implementation that actually used the underlying classes capabilities, so this was a very easy and natural move.

craigtaverner committed:
Fix failing tests by removing assertion

craigtaverner committed:
Refactor to keep methods order more consistent, and make EmptySearchStats

We found some runtime cases where there are in fact no SearchExecutionContexts, and so EmptySearchStats is a valid concept.

craigtaverner committed:
Move EmptySearchStats into static SearchStats record

craigtaverner committed:
Fix compile error

craigtaverner committed:
Remove unnecessary use of DisabledSearchStats

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_searchstats_for_indexed_docvalues

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_searchstats_for_indexed_docvalues

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_searchstats_for_indexed_docvalues

craigtaverner committed:
Wrap all lucene pushdown predicates in a single interface

This makes it much easier to differentiate spatial and non-spatial pushdown, even in complex predicates that contain combinations of both.

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_searchstats_for_indexed_docvalues

craigtaverner committed:
Move the last method out of LucenePushdownUtils, removing that class

craigtaverner committed:
Refactored tests to cover all four combinations of indexed/doc-values

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_searchstats_for_indexed_docvalues

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

craigtaverner committed:
Broaden spatial-pushdown tests to cover all combinations of indexed and doc-values

As well as a multi-index query that queries all indexes in one query to excercise the fact that different planning can happen at the SearchExecutionContext level.

costin commented on x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/EsqlTestUtils.java:
```diff
@@ -245,10 +263,76 @@ public byte[] max(String field, DataType dataType) {
+        }
+
+        @Override
+        public boolean hasIdenticalDelegate(String field) {
```
Since PR does renaming, I propose to change this to something more clearer such as "hasFallbackSubfield" / "hasExactSubfield" or something similar.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/LucenePushdownPredicates.java:
```diff
@@ -0,0 +1,111 @@
+     * For pushing down TopN and for pushing down filters with SingleValueQuery,
+     * we need to check if the field is indexed and has doc values.
+     */
+    boolean isIndexedAndDocValues(FieldAttribute attr);
```
isIndexedAndHasDocValues? :)

craigtaverner replied:
Done.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/LucenePushdownPredicates.java:
```diff
@@ -0,0 +1,111 @@
+ *         not the full-text search provided by the MATCH and QSTR functions, which are pushed down separately.</li>
+ * </ol>
+ */
+public interface LucenePushdownPredicates {
```
Do we still need the default implementation ? I wonder how out of sync it will become with respect to SearchStats.

Separately there's a lot in this class - it defines both an interface, its implementations and acts as a factory so I suggest decomposing it:
1. have FieldAttribute in the name since all methods resolve around that - e.g. LucenePushDown, LuceneFieldInfo,  etc...
2. Move the two implementation into a Xs class such as LucenePushDowns, LuceneFieldInfos,  (like we do for DataType and DataTypes) :  LucenePushDowns.DEFAULT or LucenePushDowns.from(SearchStats)

craigtaverner replied:
The recent work on MATCH has removed one of the three usages of the DEFAULT implementation. One other is test-only so could be refactored away, but the remaining use persists. It is in the `ComputeEngine.startComputeOnDataNodes` method, which creates a filter on `@timestamp` based on the current plan before we even know the separate data nodes, or individual SearchExecutionContexts to be used for creating the `SearchStats`. I'm maintaining the exact same behaviour here (ie. I did not want to change anything here, other than naming), although perhaps we would want to revisit this later. If so, we should do that in another PR.

(The above mentioned test-only usage is in `FilterTests`, but if we got rid of the ComputeEngine usage, we could mock SearchStats in FilterTests to get around that usage too, again a task for another PR if we desire it).

Regarding splitting the file into multiple files, I actually prefer the current approach because it is a simple interface with two simple implementations, and I think it is much easier to reason about with all three in the same file (which remains a very small file).

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/LucenePushdownPredicates.java:
```diff
@@ -0,0 +1,111 @@
+     * also differ from node to node, and we can physically plan each node separately, allowing Lucene pushdown on the nodes that
+     * support it, and relying on the compute engine for the nodes that do not.
+     */
+    default boolean isPushableFieldAttribute(Expression exp) {
+        if (exp instanceof FieldAttribute fa && fa.getExactInfo().hasExact() && isIndexedAndDocValues(fa)) {
+            return (fa.dataType() != DataType.TEXT && fa.dataType() != DataType.SEMANTIC_TEXT) || hasIdenticalDelegate(fa);
+        }
+        return false;
+    }
```
This method  can be moved in the LucenePushdownS (or whatever name you decide).

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/stats/SearchContextStats.java:
```diff
@@ -0,0 +1,357 @@
+        return completed ? count[0] : -1;
+    }
+
+    public byte[] min(String field, DataType dataType) {
+        var stat = cache.computeIfAbsent(field, this::makeFieldStats);
+        if (stat.min == null) {
+            var min = new byte[][] { null };
+            doWithContexts(r -> {
+                byte[] localMin = PointValues.getMinPackedValue(r, field);
+                // TODO: how to compare with the previous min
+                if (localMin != null) {
+                    if (min[0] == null) {
+                        min[0] = localMin;
+                    } else {
+                        throw new EsqlIllegalArgumentException("Don't know how to compare with previous min");
+                    }
+                }
+                return true;
+            }, true);
+            stat.min = min[0];
+        }
+        // return stat.min;
+        return null;
+    }
```
Could Ignacio help finalize these two methods (min and max)?
The min and max come handy with range queries (e.g. `WHERE x < 10`  gets folded to false if x min is 10+).

craigtaverner replied:
The `min` and `max` methods were coded in the original `SearchStats`, but as yet have never been used. I actually favour removing them, and only adding them back when they get used. A feature to fold `WHERE x < 10` to `false` on a per-shard or per collections of shards basis is certainly a nice feature, and I will discuss this with Ignacio.

costin reviewed (approved):
Left some minor comments otherwise LGTM

craigtaverner committed:
Rename some methods and parameters based on code-review

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_searchstats_for_indexed_docvalues
